File: examples/python/strategist_bot.py
"""Strategist bot: pure Layer-3 tactical decision-making.

This bot is the showcase for the toolkit's high-level layer. It subclasses
``TacticalBot`` and overrides ``decide()`` — that's all. The framework wires
up tracking, fire control, steering, sensor scheduling, and evasion. The
bot author writes *intent*, not plumbing.

Run against a local server:

    python examples/python/strategist_bot.py --host localhost --port 7878 --name strategist
"""

from __future__ import annotations

import argparse
import logging

from naval_sdk import run
from naval_sdk.tactical import Intent, PingWhenStale, TacticalBot, TacticalContext

log = logging.getLogger("strategist_bot")

LOW_HP = 30


class StrategistBot(TacticalBot):
    sensor_policy = PingWhenStale(stale_threshold_ticks=12)

    def decide(self, ctx: TacticalContext) -> Intent:
        # Hurt? Run for the corner furthest from the nearest threat.
        if ctx.me.hp < LOW_HP and ctx.threats:
            return Intent.retreat_to(self._safest_corner(ctx))

        # See a threat? Engage the closest.
        if ctx.threats:
            return Intent.engage(ctx.threats.nearest())

        # No contacts: patrol the central box to find someone.
        return Intent.patrol(
            rect=(
                ctx.map_width * 0.25,
                ctx.map_height * 0.25,
                ctx.map_width * 0.75,
                ctx.map_height * 0.75,
            )
        )

    @staticmethod
    def _safest_corner(ctx: TacticalContext):
        threat = ctx.threats.nearest()
        assert threat is not None
        margin = 60.0
        corners = [
            (margin, margin),
            (ctx.map_width - margin, margin),
            (margin, ctx.map_height - margin),
            (ctx.map_width - margin, ctx.map_height - margin),
        ]
        return max(
            corners,
            key=lambda c: (c[0] - threat.pos[0]) ** 2 + (c[1] - threat.pos[1]) ** 2,
        )


def main() -> None:
    logging.basicConfig(level=logging.INFO)
    p = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    p.add_argument("--host", default="localhost")
    p.add_argument("--port", type=int, default=7878)
    p.add_argument("--name", default="strategist")
    args = p.parse_args()
    run(StrategistBot(), host=args.host, port=args.port, name=args.name)


if __name__ == "__main__":
    main()

File: examples/java/src/main/java/com/battlesim/examples/StrategistBot.java
package com.battlesim.examples;

import com.battlesim.naval.BotRunner;
import com.battlesim.naval.protocol.Vec2;
import com.battlesim.naval.tactical.Intent;
import com.battlesim.naval.tactical.SensorPolicy;
import com.battlesim.naval.tactical.TacticalBot;
import com.battlesim.naval.tactical.TacticalContext;
import com.battlesim.naval.tactical.Track;

/**
 * Pure Layer-3 example: subclass {@link TacticalBot} and override
 * {@link #decide}. The framework wires up tracking, fire control, steering,
 * sensor scheduling, and evasion.
 */
public final class StrategistBot extends TacticalBot {
    private static final int LOW_HP = 30;
    private static final float MARGIN = 60.0f;

    public StrategistBot() {
        this.sensorPolicy = new SensorPolicy.PingWhenStale(12);
    }

    @Override
    public Intent decide(TacticalContext ctx) {
        if (ctx.me().hp() < LOW_HP && !ctx.threats().isEmpty()) {
            return Intent.retreatTo(safestCorner(ctx));
        }
        if (!ctx.threats().isEmpty()) {
            return Intent.engage(ctx.threats().nearest().orElseThrow());
        }
        return Intent.patrol(
                ctx.mapWidth() * 0.25f,
                ctx.mapHeight() * 0.25f,
                ctx.mapWidth() * 0.75f,
                ctx.mapHeight() * 0.75f);
    }

    private static Vec2 safestCorner(TacticalContext ctx) {
        Track threat = ctx.threats().nearest().orElseThrow();
        Vec2[] corners = new Vec2[] {
                new Vec2(MARGIN, MARGIN),
                new Vec2(ctx.mapWidth() - MARGIN, MARGIN),
                new Vec2(MARGIN, ctx.mapHeight() - MARGIN),
                new Vec2(ctx.mapWidth() - MARGIN, ctx.mapHeight() - MARGIN),
        };
        Vec2 best = corners[0];
        float bestSq = sq(best.x() - threat.pos().x()) + sq(best.y() - threat.pos().y());
        for (int i = 1; i < corners.length; i++) {
            float s = sq(corners[i].x() - threat.pos().x()) + sq(corners[i].y() - threat.pos().y());
            if (s > bestSq) {
                bestSq = s;
                best = corners[i];
            }
        }
        return best;
    }

    private static float sq(float v) { return v * v; }

    public static void main(String[] args) {
        BotArgs parsed = BotArgs.parse(args, "strategist");
        BotRunner.run(new StrategistBot(), parsed.host(), parsed.port(), parsed.name());
    }
}
